### ISSUE TITLE
Show blocks feature

### ISSUE THREAD

**Issue by Mgsy** [ckeditor/ckeditor5#9398]
## 📝 Provide a description of the new feature

Implement a [show blocks feature](https://ckeditor.com/cke4/addon/showblocks) that works the same way as in CKEditor 4.

---

If you'd like to see this feature implemented, add a 👍 reaction to this post.

**Comment by oleq** [comment]
Note: Both CKE4 and TinyMCE use background images to display "P", "H2", "BLOCKQUOTE" in the blocks. 

I suppose they are used instead of `::after/::before` to avoid breaking the integrator's styles. After all, content like `<p>` or `<h2>` is unlikely to have a `background-image` but it may use `::after/::before` to display some extras.

**Comment by oleq** [comment]
A solution to avoid a tedious generation of SVGs that takes advantage of our PostCSS system.

TODOs:
* final styling,
* [license of the font used by text-to-svg](https://www.npmjs.com/package/text-to-svg)

A new function mixin in `postcss-mixins`

```js
/**
 * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md.
 */

'use strict';

const TextToSVG = require( 'text-to-svg' );
const textToSVG = TextToSVG.loadSync();

/* eslint-env node */

/**
 * Returns a PostCSS configuration to build the editor styles (e.g. used by postcss-loader).
 *
 * @param {Object} options
 * @param {Boolean} options.sourceMap When true, an inline source map will be built into the output CSS.
 * @param {Boolean} options.minify When true, the output CSS will be minified.
 * @param {ThemeImporterOptions} options.themeImporter Configuration of the theme-importer PostCSS plugin.
 * See the plugin to learn more.
 * @returns {Object} A PostCSS configuration object, e.g. to be used by the postcss-loader.
 */
module.exports = function getPostCssConfig( options = {} ) {
	const config = {
		plugins: [
			require( 'postcss-import' )(),
			require( './themeimporter' )( options.themeImporter ),
			require( 'postcss-mixins' )( {
				mixins: {
					'text-to-svg-background': function( mixin, text ) {
						const attributes = { fill: 'black' };
						const options = { x: 0, y: 0, fontSize: 10, anchor: 'top', attributes: attributes };
						const svg = textToSVG.getSVG( text, options );
						const base64svg = Buffer.from( svg ).toString('base64');

						mixin.replaceWith( `background-image: url("data:image/svg+xml;base64,${ base64svg }") !important;` );
					}
				}
			} ),
			require( 'postcss-nesting' )( {
				// https://github.com/ckeditor/ckeditor5/issues/11730
				noIsPseudoSelector: true
			} ),
			require( './themelogger' )()
		]
	};

	if ( options.sourceMap ) {
		config.sourceMap = 'inline';
	}

	if ( options.minify ) {
		config.plugins.push( require( 'cssnano' )( {
			preset: 'default',
			autoprefixer: false,
			reduceIdents: false
		} ) );
	}

	return config;
};
```

and some CSS

```css
body {
	@mixin text-to-svg-background BODY;
}

.ck.ck-content p {
	@mixin text-to-svg-background PARAGRAPH;
}

.ck.ck-content figure.image figcaption {
	@mixin text-to-svg-background FIGCAPTION;
}
```

= 

<img width="884" alt="image" src="https://github.com/ckeditor/ckeditor5/assets/1099479/f2ea0e28-483c-4622-8abc-22555f563166">

**Comment by pszczesniak** [comment]
We can simplified it more. We can use SVG `<text>` so we just need to pass text that we want to show. Here is an [example](https://codepen.io/piotrekszczesniak/pen/GRYLgvN?editors=1100).

TODOs:

*   final styling/font size

Based on this @oleq made a `mixin` that just works :partying\_face: 

```postcss
@define-mixin better-text-to-svg-background $text {
	background-image: url("data:image/svg+xml;utf8,<svg width='150' height='8' xmlns='http://www.w3.org/2000/svg'><text y='8' font-size='10px' font-family='monospace'>$(text)</text></svg>");
	background-repeat: no-repeat;
	background-position: 2px 2px;
	padding: 10px;
	border: 1px dashed #000;
}

.ck.ck-content ol, .ck.ck-content ul {
	@mixin better-text-to-svg-background LIST;
}

.ck.ck-content blockquote {
	@mixin better-text-to-svg-background BLOCKQUOTE;
}

.ck.ck-content p {
	@mixin better-text-to-svg-background PARAGRAPH;
}

.ck.ck-content figure.image figcaption {
	@mixin better-text-to-svg-background FIGCAPTION;
}
```

And here is the result:

![image (1)](https://github.com/ckeditor/ckeditor5/assets/7074833/55d4be8f-d7e5-470a-9fbd-4313f7fc6286)

**Comment by oleq** [comment]
That should save us a lot of work 🙂

**Comment by Dumluregn** [comment]
Epic branch name: `ck/epic/9398-show-blocks`.

**Comment by Dumluregn** [comment]
Feature is available since 38.1.0: https://ckeditor.com/docs/ckeditor5/latest/features/show-blocks.html :tada:
